### PR TITLE
Fast parallel coordinates and alpha 

### LINKED ISSUE #447: Fix ParallelCoordinates tests and add fast parameter
See #420 

- [x] fix image comparison tests for parallel coordinates 
- [x] add `fast=False` argument and docstring 
- [x] create fast and regular drawing methods based on parameter 
- [x] add section in documentation explaining fast vs. slow 
- [x] update #230 and #59

### PR DESCRIPTION
**Pull request by bbengfort**
Fixes #447 and #59  

Implements `alpha` and `fast` parameters in `ParallelCoordinates`, updates the documentation with the difference between fast and slow parallel coordinates and creates separate methods for drawing by instance and drawing by class. 

This PR is currently a work in progress and does not yet need to be reviewed. 

TODO: 

- [x] implement drawing methods 
- [x] write image comparison tests for fast vs. slow 
- [x] test dataframe support for parallel coordinates
- [x] test alpha support in parallel coordinates  
- [x] finish documentation update
- [x] create new issue for "Create an optimization technique that reorders the columns such that the overlap of two instances by the same class is minimized" so that #59 can be closed.  

@thekylesaurus @rebeccabilbro

### CODE REVIEW

**Comment by bbengfort** [comment]
@thekylesaurus I've completed my suggested updates; would you mind code reviewing with us?

**Comment by bbengfort** [comment]
See #452

File: yellowbrick/features/pcoords.py [hunk 1]
```diff
@@ -1,28 +1,29 @@
...
+# Created: Mon Oct 03 21:46:06 2016 -0400
 #
 # Copyright (C) 2016 District Data Labs
 # For license information, see LICENSE.txt
 #
 # ID: pcoords.py [0f4b236] benjamin@bengfort.com $
 
 """
-Implementations of parallel coordinates for multi-dimensional feature
-analysis. There are a variety of parallel coordinates from Andrews Curves to
```
Comment (rebeccabilbro) [inline yellowbrick/features/pcoords.py]: Reminder to check to see if we still have an open issue to implement Andrews Curves as an independent visualizer, and if so, consider the extent to which these fast/slow issues may also be relevant there.

File: yellowbrick/features/pcoords.py [hunk 2]
```diff
@@ -87,6 +89,17 @@ def parallel_coordinates(X, y, ax=None, features=None, classes=None,
...
+    alpha : float, default: None
+        Specify a transparency where 1 is completely opaque and 0 is completely
+        transparent. This property makes densely clustered lines more visible.
+        If None, the alpha is set to 0.5 in "fast" mode and 0.25 otherwise.
+
+    fast : bool, default: False
+        Fast mode improves the performance of the drawing time of parallel
+        coordinates but produces an image that does not show the overlap of
+        instances in the same class. Fast mode should be used when drawing all
+        instances is too burdensome and sampling is not an option.
```
Comment (rebeccabilbro) [inline yellowbrick/features/pcoords.py]: Thanks for the clear docstring explanation!

File: yellowbrick/features/pcoords.py [hunk 3]
```diff
@@ -328,51 +392,100 @@ def draw(self, X, y, **kwargs):
...
 
-        # Get shape of the sampled data
-        nrows, ncols = X.shape
+    def draw_instances(self, X, y, **kwargs):
+        """
+        Draw the instances colored by the target y such that each line is a
+        single instance. This is the "slow" mode of drawing, since each
+        instance has to be drawn individually. However, in so doing, the
+        density of instances in braids is more apparent since lines have an
+        independent alpha that is compounded in the figure.
```
Comment (rebeccabilbro) [inline yellowbrick/features/pcoords.py]: Is there a way to numerically represent braid density (a la silhouettes) that could be computed and added to the figure?

File: yellowbrick/features/pcoords.py [hunk 4]
```diff
@@ -328,51 +392,100 @@ def draw(self, X, y, **kwargs):
...
+
+        return self.ax
+
+    def draw_classes(self, X, y, **kwargs):
+        """
+        Draw the instances colored by the target y such that each line is a
+        single class. This is the "fast" mode of drawing, since the number of
+        lines drawn equals the number of classes, rather than the number of
+        instances. However, this drawing method sacrifices inter-class density
+        of points using the alpha parameter.
```
Comment (rebeccabilbro) [inline yellowbrick/features/pcoords.py]: Nice description of the trade-off here.

File: tests/test_features/test_pcoords.py [hunk 5]
```diff
@@ -17,134 +18,279 @@
...
+
+    def test_normalized_minmax_fast(self):
+        """
+        Test image closeness on minmax normalized 3 class dataset in fast mode
+        """
+        visualizer = ParallelCoordinates(normalize='minmax', fast=True)
+        visualizer.fit_transform(self.dataset.X, self.dataset.y)
+        visualizer.poof()
+        self.assert_images_similar(visualizer, tol=0.25)
+
```
Comment (rebeccabilbro) [inline tests/test_features/test_pcoords.py]: Maybe this is overkill, but is it worth it to test that "fast" is at least as fast as "slow" given the same dataset?

File: docs/api/features/pcoords.rst [hunk 6]
```diff
@@ -51,15 +54,36 @@ Additionally the domain of each feature may make the visualization hard to inter
...
 
+Now we can see that each feature is in the range ``[-3, 3]`` where the mean of the feature is set to zero and each feature has a unit variance applied between ``[-1, 1]`` (because we're using the ``StandardScaler`` via the ``standard`` normalize parameter). This version of parallel coordinates gives us a much better sense of the distribution of the features and if any features are highly variable with respect to any one class.
+
+
+Faster Parallel Coordinates
+---------------------------
+
+Parallel coordinates can take a long time to draw since each instance is represented by a line for each feature. Worse, this time is not well spent since a lot of overlap in the visualization makes the parallel coordinates less understandable. We propose two solutions to this:
+
+1. Use ``sample=0.2`` and ``shufle=True`` parameters to shuffle and sample the dataset being drawn on the figure. The sample parameter will perform a uniform random sample of the data, selecting the percent specified.
```
Comment (rebeccabilbro) [inline docs/api/features/pcoords.rst]: should be ``shuffle=True``

File: docs/api/features/pcoords.rst [hunk 7]
```diff
@@ -3,46 +3,49 @@
 Parallel Coordinates
 ====================
 
-Parallel coordinates displays each feature as a vertical axis spaced
-evenly along the horizontal, and each instance as a line drawn between
-each individual axis. This allows many dimensions; in fact given
-infinite horizontal space (e.g. a scrollbar), an infinite number of
-dimensions can be displayed!
+Parallel coordinates is multi-dimensional feature visualization technique  where the vertical axis is duplicated horizontally for each feature. Instances are displayed as a single line segment drawn from each vertical axes to the location representing their value for that feature. This allows many dimensions to be visualized at once; in fact given infinite horizontal space (e.g. a scrolling window), technically an infinite number of dimensions can be displayed!
```
Comment (rebeccabilbro) [inline docs/api/features/pcoords.rst]: Really nice fleshing out of the user perspective here.

File: docs/api/features/pcoords.py [hunk 8]
```diff
@@ -10,30 +28,102 @@ def pcoords(X, y, outpath, **kwargs):
...
+        oz.finalize()
+
+    plt.tight_layout()
+    plt.savefig("images/fast_vs_slow_parallel_coordinates.png")
 
-    # Extract the numpy arrays from the data frame
-    X = data[features].as_matrix()
-    y = data.occupancy.as_matrix()
+
+def plot_speedup(trials=5, factors=np.arange(1, 11)):
```
Comment (rebeccabilbro) [inline docs/api/features/pcoords.py]: Love this demo! Has a very sklearn feel :)

File: yellowbrick/features/pcoords.py (hunk 1 above)
Comment (bbengfort) [reply yellowbrick/features/pcoords.py]: Yep - #100; I'll add a comment about fast and slow referencing this PR.

File: yellowbrick/features/pcoords.py (hunk 3 above)
Comment (bbengfort) [reply yellowbrick/features/pcoords.py]: Great idea, referred to in #452

File: tests/test_features/test_pcoords.py (hunk 5 above)
Comment (bbengfort) [reply tests/test_features/test_pcoords.py]: This would be tough to do in our current test framework, we'd have to implement some benchmarking tools;  but you're right - probably as a result of this suite of issues it is something we should look into. Have added #460 in response.

**Comment by thekylesaurus** [comment]
Hi all - I'm sorry I ghosted on this project so quickly after PyCon!  Very predictably, work swallowed me back it.  It is extremely cool to see this work get implemented, however.  Thanks @bbengfort for grabbing the baton.

As soon as I am on more stable ground, I plan to become more active with the yellowbrick project.

Thanks!
